### DISCUSSION TITLE
[Nette] About migrating 51K lines of code to Laravel

### DISCUSSION THREAD

**Discussion by genesiscz** [rectorphp/rector#5002] [General]
Hello. I tried to create a discussion thread but couldn't because it is not enabled (or whatever).

I am in process of refactoring a big Nette project using no Doctrine (only NetteDatabase) (51k lines of code) and preparing it to be migrated to Laravel. 

Many things were done by using generators (such as DB Entities), but many of these things will probably need help from Rector. I know I probably have to create a new set for that, and would love your help. 

Are there any sets I can use as a whole, or at least point me to some specific Rector Rules that comes into your mind?

I know I will specifically need RenameClassRector, RenameMethodRector.

I for sure know I will need rule to transfer all parameters from action*, render*

- ```render(int $id, ?int $param1, ?string $param2, int $param3)``` -> ```render(Request $request)```
- but then in the first line ```[$id, $param1, $param2, $param3] = $request->getParameters(["id", "param1", "param2", "param3"]```
If you had time to help me with this one, even in pseudocode/description, it would be really appreciated (and for sure donated if that saves me even 10 hours of my work)

**Comment by TomasVotruba** [comment]
Hi,

I've moved it here to discussion, as it's the place you wanted at first.

Congrats on a honorable challange! I think many project would benefit from such migration set.
I don't have any experience with Nette to Laravel. We did another migration, that is not the same, but will have very similar procedure. We migrated Nette to Symfony, you can read about it here:

- https://tomasvotruba.com/blog/2019/02/21/how-we-migrated-from-nette-to-symfony-in-3-weeks-part-1/
- https://tomasvotruba.com/blog/2019/03/07/why-we-migrated-from-nette-to-symfony-in-3-weeks-part-2/
- https://tomasvotruba.com/blog/2019/03/11/why-we-migrated-from-nette-to-symfony-in-3-weeks-part-3/

There is lot of generic knowledge about framework migrations, that would save your some blind branches.

The set can be found here: https://github.com/rectorphp/rector/blob/master/config/set/nette-to-symfony.php

<br>

**Personally, I'd just take it and modify custom rules from Symfony to Laravel. They're very alike (Laravel build on Symfony ideology), so that would only syntax tweaks. Whole work could be matter of 1 week**. Depends what mess is the Nette project in :) 

If you want to speed-up learning process, I can offer you private consulting :+1:

**Reply by genesiscz** [reply]
I am gonna dive in and consider if I need a private consulting. What is your hourly salary?

I was actually inspired by these articles to do this migration in the first place. It's pretty messy but is getting better. :-)

Another thing is that we're moving from templates into API (as we are going to use React for the whole thing). So everything we have thrown into the templates now has to be carefully moved out into JSON response without exposing too much. 

Another question came into mind - what do you use to document REST API?

**Reply by TomasVotruba** [reply]
Depends on the project and time spent, around 90-120 €. Contact me on email for details.

> It's pretty messy but is getting better. :-)

Yeah, first migration is the worst :) but second feel more like copy-pasting.

So migrating templates to json + React? That sounds cool. But, in my experience people can easily get stuck on doing many migrations at once. I recommend doing the framework first. Completely. Then doing anything else.

**Comment by TomasVotruba** [comment]
I'm curious, how is it going? :)

**Reply by genesiscz** [reply]
I created a first Rule with

```php
	public function getDefinition(): RectorDefinition
	{
		return new RectorDefinition(
			'Move action/render methods parameters inside', [
			new CodeSample(
				<<<'CODE_SAMPLE'
final class SomeClass
{
    public function actionDetail(int $id, ?string $date = null)
    {

    }
}
CODE_SAMPLE
				,
				<<<'CODE_SAMPLE'
final class SomeClass
{
    public function run(Request $request)
    {
        [$id, $date] = $request->getParameters(["id", "date"]); 
    }
}
CODE_SAMPLE
			),
		]
		);
	}
```

and another one

```php
	public function getDefinition(): RectorDefinition
	{
		return new RectorDefinition(
			'Create DocBlocks for repositories', [
			new CodeSample(
				<<<'CODE_SAMPLE'
final class Drivers extends BaseRepository
{

}
CODE_SAMPLE
				,
				<<<'CODE_SAMPLE'
/**
 * @method Driver|null get(int $id)
 * @method Driver|EloquentBuilder getTable()
 * @method Driver|null last(EloquentBuilder $selection)
 * @method Driver|null firstWhere($condition, ...$params)
 * @method Driver|EloquentBuilder table()
 * @template TModelClass of Driver
 */
final class Drivers extends BaseRepository
{

}
CODE_SAMPLE
			),
		]
		);
	}
```

That's pretty much it for now, as the most problematic part is Eloquent static magic (Models are forwarding where() and many other calls into underlying Builder by __callStatic) and I haven't figured out how to cleanly translate all `$this->userRepository->where()->order()->group()->fetch()` calls into something that wouldn't totally kill all the repositories. I don't want to use `User::where()->order()->group()->fetch()` as is typical for Laravel, but the more i dig in the more problematic it goes. 

This is one obstacle that I've been trying to figure out for a while and didn't find an elegant solution for this. If you have any idea how to do this right, it'd be appreciated.

**Reply by genesiscz** [reply]
On another note, I am preparing the code base to replace most of the ActiveRow, Selection mess. Rector helps a lot with ActiveRows, as I replaced most of the underlying Repositories to return specific model (I am still not sure if Rector respects generic types of PHPStan such as @return TModelClass because it works in most instances, but not everywhere).

Many things were done manually because they were pretty easy to do with basic string/regex replacing, such as 

```
  $this->driverRepository->where("registered_at BETWEEN ? AND ?", $start, $stop);
```

into

```
  $this->driverRepository->whereBetween("registered_at", [$start, $stop]);
```

I am not sure it makes a lot of sense, because this function forwards the call directly to Model which forwards it magically directly to EloquentBuilder. It would make more sense to move the logic into dedicated EloquentBuilders, because they give me a fluent syntax to work with.

**Reply by TomasVotruba** [reply]
Looks good. Could you send those ad PR in Rector?
I'll be able to help then

**Reply by genesiscz** [reply]
I am not sure how would the Repository docBlocks Rector help the general public. 
The parameters extraction is more possible to be useful for somebody, but probably requires a little bit of manual tuning, too, which is, I guess, not welcome. I can post a PR of the second one, but after I actually apply the Rector without much errors.

**Reply by TomasVotruba** [reply]
The 1st and 2nd rule are generic for Nette and Laravel, so they would save time or next similar migrations.
I can polish the manual tuning to make them generic, don't worry about it.

**Reply by genesiscz** [reply]
What if the rule uses methods from Laravel Support package, such as `$modelName = Str::singular($repositoryClass);`?

**Reply by TomasVotruba** [reply]
If it's not pulling whole Laravel, it should be fine :) 

If it is, we can use this package instead: https://symfony.com/blog/new-in-symfony-5-1-deprecated-the-inflector-component

**Reply by genesiscz** [reply]
Is it okay if it fails for some cases then? 

```
dump(Str::singular("ApiDevices")); // -> ApiDevice
dump($inflector->singularize("ApiDevices")); -> ["ApiDevex", "ApiDevix", "ApiDevice"]
```

-----

On another note, I've created a third Rule to add/remove null type to model properties which are nullable based on information_schema. How would you go with database query/connection right from the Rector Rule?

```patch
    ---------- begin diff ----------
--- Original
+++ New
@@ -7,15 +7,15 @@
 use Illuminate\Database\Eloquent\Relations\HasMany;
 /**
  * @property int $id
- * @property int $race_result_id
- * @property int $driver_id
- * @property int $race_id
+ * @property int|null $race_result_id
+ * @property int|null $driver_id
+ * @property int|null $race_id
  * @property int $position
  * @property int $best_lap_time
  * @property int $timestamp
- * @property Race $race
- * @property Driver $driver
- * @property RaceResult $raceResult
+ * @property Race|null $race
+ * @property Driver|null $driver
+ * @property RaceResult|null $raceResult
  * @property RaceResultDriverBadge[]|Collection $raceResultDriverBadges
  */
 class RaceResultDriver extends BaseModel
    ----------- end diff -----------
```

**Comment by genesiscz** [comment]
I am currently struggling with the function params into Request $request Rector rule. 

```php

	/**
	 * @param ClassMethod $classMethod
	 */
	private function migrateToFunction(ClassMethod $classMethod): ?ClassMethod
	{
		$args = [];

		/** @var Param $param */
		foreach ($classMethod->getParams() as $param) {
			// function actionTest(int $id, int $id2)
			$args[] = $this->getName($param);
		}

		// first line of function
		$first = $classMethod->getStmts()[0];
		if (!$first) {
			return null;
		}

		foreach($args as $var) {
			// $request->input('id');
			$methodCall = $this->createMethodCall(
				'request', "input", [
					$this->createArg($var)
				]
			);

			$assign = new Assign(new Variable($var), $methodCall);
			$this->addNodeBeforeNode($assign, $first);
		}

		$classMethod->params = [
			new Param(new Variable("request"), null, new FullyQualified(\App\Requests\Request::class))
		];

		return $classMethod;
	}
```

Any idea why would this remove lines between other statements inside of that method?

```patch
    ---------- begin diff ----------
--- Original
+++ New
@@ -72,22 +72,22 @@
                }
        }

-       function actionDefault($reservation_id): void
+       function actionDefault(\App\Requests\Request $request): void
        {
+               $reservation_id = $request->input('reservation_id');
                $this->redirect("hall", ["id" => 1]);
        }

-       function actionHall($id, $reservation_id): void
+       function actionHall(\App\Requests\Request $request): void
        {
+               $id = $request->input('id');
+               $reservation_id = $request->input('reservation_id');
                $this->template->hall = $hall = $this->getHall($id);
-
                $reservations_q = $this->hall_reservations->where("hall_id", $id);
                $reservations = [];
-
                foreach ($reservations_q as $reservation) {
                        $reservations[$reservation->hour][$reservation->minute] = $reservation;
                }
-
                $this->template->reservations = $reservations;
                $this->template->reservation_id = $reservation_id;
        }
```
